crates/sui-kvstore/src/lib.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+mod bigtable;
+use anyhow::Result;
+use async_trait::async_trait;
+pub use bigtable::client::BigTableClient;
+use sui_types::base_types::SequenceNumber;
+use sui_types::digests::TransactionDigest;
+use sui_types::effects::{TransactionEffects, TransactionEvents};
+use sui_types::full_checkpoint_content::CheckpointData;
+use sui_types::messages_checkpoint::{
+    CheckpointContents, CheckpointSequenceNumber, CheckpointSignatureMessage, CheckpointSummary,
+};
+use sui_types::object::Object;
+use sui_types::storage::ObjectKey;
+use sui_types::transaction::Transaction;
+
+#[async_trait]
+pub trait KeyValueStoreReader {
```

Comment:
Noting the desire to have a way to query for an object that's less than or equal to a particular version.

If it's not possible to serve that from a key-value store easily, then that's okay, we can handle it on the indexer side, but I wanted to make sure we'd explored this option.

Reply:
It should be possible with the existing schema in BigTable. We can add it on top. I was a bit conservative with the API so we could support the same functionality for different backends, and not all of them can serve/represent such queries. But maybe it could be optional functionality that some backends support and some don't

Reply:
Okay good to know -- it would be good to have a sketch of what that looks like. 

I realised yesterday that the full query we need to support is "give me the latest version of object with ID `O`, whose version is less than or equal to `V`, at checkpoint `C`" (there's an extra checkpoint component here).

Today, we implement that using a dedicated table of object versions and checkpoints in postgres, and it would be okay to keep doing it that way and offer this query with e.g. a couple of years of retention. It supports accessing dynamic fields.

Reply:
I think it makes sense and is easy to allow for bounding by version, less so by checkpoint but i'd need to think on it a bit more

Reply:
Oh, I see. Unfortunately, the current schema won't have support for such query, specifically the at checkpoint C part; it only allows to fetch the latest version available in the KV store. 
Theoretically, it could be done in BigTable by replacing multi-get with multi-scans backwards and an extra fetch if accumulated rows are from a checkpoint `> target_checkpoint`. 
But this approach could lead to inefficient queries, and I'm not sure we could easily support that in other backends. 
Overall, I don't think KV is the right place for this functionality.